llama_index/indices/query/embedding_utils.py

```diff
@@ -92,3 +92,62 @@ def get_top_k_embeddings_learner(
     result_ids = [embedding_ids[ix] for ix in top_sorted_ix]
 
     return result_similarities, result_ids
+
+
+def get_top_k_mmr_embeddings(
```

Comment:
Similarity will go as O(n log(n)), because you're adding elements to the heap, whereas MMR goes as O(n^2 log(n)), as each result could affect the other results and there's a re-sort at every step. It's an iterative algorithm. I agree that performance could suffer at scale.

What is best practice for llama index around benchmarking for scale? I could add a test that does a larger computation. Another alternative would be to add a notebook with an example using a bigger dataset.

Reply:
Ran a test over a large number of examples (embedding size 4096, embeddings 200, top_k 100 and I'm seeing a 75x slowdown, I'll see if I can fix it up.

Reply:
Thanks for checking! 🙏

Reply:
Brought back the similarity heap, doesn't look like it improves anything. As one needs to iterate through the heap and re-heapify, that was kind of expected. I'll look into more numpy-intensive approaches and see if that can help.

Reply:
Very tricky problem due to the sorts on every loop I think 🤔 But tbh, a typical use-case might be something more like a top-k of 3, picking from 1,000ish embeddings. When I benchmarked in the past I just hard-coded the top k and scaled the embeddings lol

Reply:
I took a look at it and did some profiling. The similarity_fn, which makes a call to compare two embeddings, take up 95-99% of the compute time. For example, I'd tested an example with an embedding size of 50, 5000 embeddings, and similarity_top_k of 50. Overall, MMR is a slower algorithm than just simple similarity to the query vector, it adds additional compute steps. There could be ways to optimize past it by cutting down on the number of embedding-embedding calculations, but that goes beyond the scope of iterative MMR described in the '98 paper.

I did clean up some of the logic and remove the extra sorts. Rather than maintaining a list of results each loop, it just tracks the best result, which avoids it needing to sort. This drops the runtime down to O(n^2).

Reply:
It's just 5~10 seconds for top-100 for 1000s of embeddings of large dimensionality. So it's slower but manageable for the use cases you described.

Reply:
Thanks for taking the time to document and profile this, it helps tons! I think this should be in a good spot to merge now 💪💪